src/library_manager/lib_manager.c

```diff
@@ -611,89 +615,124 @@ int lib_manager_load_library(uint32_t dma_id, uint32_t lib_id)
 		.block_count = 1,
 		.head_block = &dma_block_cfg,
 	};
-	struct lib_manager_dma_ext dma_ext;
-	uint32_t addr_align;
-	int ret, ret2;
-	void __sparse_cache *man_tmp_buffer = NULL;
+	int ret;
 
-	if (!lib_id || lib_id >= LIB_MANAGER_MAX_LIBS) {
-		tr_err(&lib_manager_tr,
-		       "lib_manager_load_library(): invalid lib_id: %u", lib_id);
-		return -EINVAL;
-	}
+	if (_ext_lib->runtime_data)
+		return 0;
 
-	lib_manager_init();
+	dma_ext = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			  sizeof(*dma_ext));
+	if (!dma_ext)
+		return -ENOMEM;
 
-	ret = lib_manager_dma_init(&dma_ext, dma_id);
+	ret = lib_manager_dma_init(dma_ext, dma_id);
 	if (ret < 0)
-		return ret;
+		goto err_dma_init;
 
-#if CONFIG_ZEPHYR_NATIVE_DRIVERS
-	ret = dma_get_attribute(dma_ext.dma->z_dev, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
-				&addr_align);
-#else
-	ret = dma_get_attribute_legacy(dma_ext.dma, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
-				       &addr_align);
-#endif
+	ret = dma_get_attribute(dma_ext->dma->z_dev, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
+				&dma_ext->addr_align);
 	if (ret < 0)
-		goto e_init;
+		goto err_dma_init;
 
-	/* allocate temporary manifest buffer */
-	man_tmp_buffer = (__sparse_force void __sparse_cache *)rballoc_align(0, SOF_MEM_CAPS_DMA,
-								MAN_MAX_SIZE_V1_8, addr_align);
-	if (!man_tmp_buffer) {
-		ret = -ENOMEM;
-		goto e_init;
-	}
-
-	ret = lib_manager_dma_buffer_alloc(&dma_ext, MAN_MAX_SIZE_V1_8, addr_align);
+	ret = lib_manager_dma_buffer_alloc(dma_ext, MAN_MAX_SIZE_V1_8);
 	if (ret < 0)
-		goto e_buf;
+		goto err_dma_buffer;
+
+	dma_block_cfg.dest_address = dma_ext->dma_addr;
 
 	/*
 	 * make sure that the DSP is running full speed for the duration of
 	 * library loading
 	 */
 	ret = core_kcps_adjust(cpu_get_id(), CLK_MAX_CPU_HZ / 1000);
 	if (ret < 0)
-		goto cleanup;
+		goto err_dma_buffer;
 
-	dma_block_cfg.dest_address = dma_ext.dma_addr;
-
-	ret = dma_config(dma_ext.chan->dma->z_dev, dma_ext.chan->index, &config);
+	ret = dma_config(dma_ext->chan->dma->z_dev, dma_ext->chan->index, &config);
 	if (ret < 0)
-		goto kcps_rollback;
+		goto err_dma;
 
-	ret = dma_start(dma_ext.chan->dma->z_dev, dma_ext.chan->index);
+	ret = dma_start(dma_ext->chan->dma->z_dev, dma_ext->chan->index);
 	if (ret < 0)
+		goto err_dma;
+
+	_ext_lib->runtime_data = dma_ext;
+
+	return 0;
+
+err_dma:
+	core_kcps_adjust(cpu_get_id(), -(CLK_MAX_CPU_HZ / 1000));
+
+err_dma_buffer:
+	lib_manager_dma_deinit(dma_ext, dma_id);
+
+err_dma_init:
+	rfree(dma_ext);
+	return ret;
+}
+
+int lib_manager_load_library(uint32_t dma_id, uint32_t lib_id, uint32_t type)
+{
+	void __sparse_cache *man_tmp_buffer;
+	struct lib_manager_dma_ext *dma_ext;
+	struct ext_library *_ext_lib;
+	int ret, ret2;
+
+	if ((lib_id >= LIB_MANAGER_MAX_LIBS) ||
+	    (type == SOF_IPC4_GLB_LOAD_LIBRARY && lib_id == 0)) {
```

Comment:
Or to put the other way: I would have checked the lib_id validity only in case of LOAD_LIBRARY message, again to eb precise that it only used in that case.